### ISSUE TITLE
State evolution - can not make existing Int field nullable and add a new mandatory field at the same time

### ISSUE THREAD

**Issue by marko-mets-gt** [corda/corda#5649]
**Version**: Corda 4.1 OS
When having an existing state that has a mandatory Int parameter then it is impossible to make this parameter nullable and add a new non-nullable parameter at the same time.
* Stacktrace
```
Cannot construct evolution serializer for remote type net.corda.serialization.internal.amqp.EvolvabilityTests$addMandatoryFieldWithAltConstructorAndMakeExistingIntFieldNullable$CC: Mandatory property b of local type is not present in remote type. This implies the type has not evolved in a backwards compatible way. Consider making b nullable in the newer version of this type.

      Full type information:
      net.corda.serialization.internal.amqp.EvolvabilityTests$addMandatoryFieldWithAltConstructorAndMakeExistingIntFieldNullable$CC
a: int
net.corda.serialization.internal.amqp.EvolutionSerializationException:       Cannot construct evolution serializer for remote type net.corda.serialization.internal.amqp.EvolvabilityTests$addMandatoryFieldWithAltConstructorAndMakeExistingIntFieldNullable$CC: Mandatory property b of local type is not present in remote type. This implies the type has not evolved in a backwards compatible way. Consider making b nullable in the newer version of this type.

      Full type information:
      net.corda.serialization.internal.amqp.EvolvabilityTests$addMandatoryFieldWithAltConstructorAndMakeExistingIntFieldNullable$CC
a: int
	at net.corda.serialization.internal.amqp.DefaultEvolutionSerializerFactory.validateEvolvability(EvolutionSerializerFactory.kt:111)
	at net.corda.serialization.internal.amqp.DefaultEvolutionSerializerFactory.getEvolutionSerializer(EvolutionSerializerFactory.kt:66)
	at net.corda.serialization.internal.amqp.DefaultEvolutionSerializerFactory.getEvolutionSerializer(EvolutionSerializerFactory.kt:41)
	at net.corda.serialization.internal.amqp.DefaultRemoteSerializerFactory.getUncached(RemoteSerializerFactory.kt:107)
	at net.corda.serialization.internal.amqp.DefaultRemoteSerializerFactory.access$getUncached(RemoteSerializerFactory.kt:48)
	at net.corda.serialization.internal.amqp.DefaultRemoteSerializerFactory$get$1$$special$$inlined$mapValues$lambda$1.invoke(RemoteSerializerFactory.kt:78)
	at net.corda.serialization.internal.amqp.DefaultRemoteSerializerFactory$get$1$$special$$inlined$mapValues$lambda$1.invoke(RemoteSerializerFactory.kt:48)
	at net.corda.serialization.internal.amqp.DefaultDescriptorBasedSerializerRegistry.getOrBuild(DescriptorBasedSerializerRegistry.kt:28)
	at net.corda.serialization.internal.amqp.DefaultRemoteSerializerFactory$get$1.invoke(RemoteSerializerFactory.kt:77)
	at net.corda.serialization.internal.amqp.DefaultRemoteSerializerFactory$get$1.invoke(RemoteSerializerFactory.kt:48)
	at net.corda.serialization.internal.amqp.DefaultDescriptorBasedSerializerRegistry.getOrBuild(DescriptorBasedSerializerRegistry.kt:28)
	at net.corda.serialization.internal.amqp.DefaultRemoteSerializerFactory.get(RemoteSerializerFactory.kt:67)
	at net.corda.serialization.internal.amqp.ComposedSerializerFactory.get(SerializerFactory.kt)
	at net.corda.serialization.internal.amqp.DeserializationInput.readObject$serialization(DeserializationInput.kt:172)
	at net.corda.serialization.internal.amqp.DeserializationInput.readObjectOrNull$serialization(DeserializationInput.kt:147)
	at net.corda.serialization.internal.amqp.DeserializationInput$deserialize$1.invoke(DeserializationInput.kt:124)
	at net.corda.serialization.internal.amqp.DeserializationInput.des(DeserializationInput.kt:99)
	at net.corda.serialization.internal.amqp.DeserializationInput.deserialize(DeserializationInput.kt:119)
```
* Steps to reproduce the issue
Unit test, run it from `EvolvabilityTests.kt`:
```
@Test
fun addMandatoryFieldWithAltConstructorAndMakeExistingIntFieldNullable() {
    val sf = testDefaultFactory()
    val A = 1
    val resource = "EvolvabilityTests.addMandatoryFieldWithAltConstructorAndMakeExistingIntFieldNullable"
    // Original version of the class as it was serialised
    // data class CC(val a: Int)
    // File(URI("$localPath/$resource")).writeBytes(SerializationOutput(sf).serialize(CC(A)).bytes)
    @Suppress("UNUSED")
    data class CC(val a: Int?, val b: String) {
        @DeprecatedConstructorForDeserialization(1)
        constructor (a: Int) : this(a, "42")
    }
    val url = EvolvabilityTests::class.java.getResource(resource)
    val sc2 = url.readBytes()
    val deserializedCC = DeserializationInput(sf).deserialize(SerializedBytes<CC>(sc2))
    assertEquals(A, deserializedCC.a)
    assertEquals("42", deserializedCC.b)
}
```
NB! If parameter `b` would be a nullable string (`val b: String?`) then this test would fail on last assertion (`assertEquals("42", deserializedCC.b)`) as `deserializedCC.b` would be `null`.

**Comment by amosmwsmith** [comment]
Hi Marko,
The problem here (courtesy of R3 Engineering team) is that making a nullable changes its underlying property type from int to java.lang.Integer. The Evolution Serializer does not realise that int can still be assigned to java.lang.Integer once it has been “boxed”, and so EvolutionSerializerFactory.findEvolverConstrutor() does not recognise the annotated constructor as being suitable.

With the annotated constructor excluded, the serialization code is forced to look at the primary constructor instead, and cannot satisfy its new mandatory b parameter.

A PR has been made by Chris Rankin in 4.4 version already and is pending merge - the fix is as follows:

```
private fun findEvolverConstructor(constructors: List<EvolutionConstructorInformation>,
                                       properties: Map<String, RemotePropertyInformation>): EvolutionConstructorInformation? {
        val propertyTypes = properties.mapValues { (_, info) -> info.type.typeIdentifier.getLocalType(classLoader).asClass() }
// Evolver constructors are listed in ascending version order, so we just want the last that matches.
        return constructors.lastOrNull { (_, evolverProperties) ->
            // We have a match if all mandatory evolver properties have a type-compatible property in the remote type.
            evolverProperties.all { (name, evolverProperty) ->
                val propertyType = propertyTypes[name]
                if (propertyType == null) !evolverProperty.isMandatory
                else {
                    val evolverPropertyType = evolverProperty.type.observedType.asClass()
                    evolverPropertyType.isAssignableFrom(propertyType) ||
                            (propertyType.isPrimitive && evolverPropertyType.isAssignableFrom(propertyType.kotlin.javaObjectType))
                }
            }
        }
    }
```

**Comment by davidrapacchiale** [comment]
@chrisr3 can you please link the PR that fixes it and close this issue?

**Comment by chrisr3** [comment]
This issue is fixed by PR #5661 .
